extended/src/main/java/io/kubernetes/client/extended/kubectl/KubectlPortForward.java

```diff
@@ -78,47 +78,44 @@ public void shutdown() {
   private void executeInternal()
       throws ApiException, KubectlException, IOException, InterruptedException {
     PortForward pf = new PortForward(apiClient);
-    PortForward.PortForwardResult result = pf.forward(namespace, name, targetPorts);
-    if (result == null) {
-      throw new KubectlException("PortForward failed!");
-    }
     // TODO: Convert this to NIO to reduce the number of threads?
     List<Thread> threads = new ArrayList<>();
     for (int i = 0; i < localPorts.size(); i++) {
       int targetPort = targetPorts.get(i);
       threads.add(
-          portForward(
-              new ServerSocket(localPorts.get(i)),
-              result.getInputStream(targetPort),
-              result.getOutboundStream(targetPort)));
+          portForward(pf,
+              new ServerSocket(localPorts.get(i)), targetPort));
     }
     for (Thread t : threads) {
       t.join();
     }
   }
 
-  private Thread portForward(ServerSocket server, InputStream in, OutputStream out) {
-    Thread t =
-        new Thread(
-            new Runnable() {
-              @Override
-              public void run() {
-                while (running) {
-                  try {
-                    Socket sock = server.accept();
-                    Thread t1 = copyAsync(sock.getInputStream(), out, onUnhandledError);
-                    Thread t2 = copyAsync(in, sock.getOutputStream(), onUnhandledError);
-
-                    t1.join();
-                    t2.join();
-                  } catch (InterruptedException | IOException ex) {
-                    Optional.ofNullable(onUnhandledError)
-                        .orElse(Throwable::printStackTrace)
-                        .accept(ex);
-                  }
+  private Thread portForward(PortForward pf, ServerSocket server, int targetPort) {
+    Thread t = new Thread(
+        new Runnable() {
+          @Override
+          public void run() {
+            while (running) {
+              try (Socket sock = server.accept()) {
+                PortForward.PortForwardResult result = pf.forward(namespace, name, Arrays.asList(targetPort));
+                if (result == null) {
+                  throw new KubectlException("PortForward failed!");
                 }
+                InputStream in = result.getInputStream(targetPort);
+                OutputStream out = result.getOutboundStream(targetPort);
+                Thread t1 = copyAsync(sock.getInputStream(), out, onUnhandledError);
+                Thread t2 = copyAsync(in, sock.getOutputStream(), onUnhandledError);
+
+                t1.join();
+                in.close();
+                t2.join();
+              } catch (Exception ex) {
+                onUnhandledError.accept(ex);
```

Comment:
`onUnhandledError` can be null (I think) so can you retain the `Optional` code from the original?

Reply:
It can only be null if the user intentionally sets it to null, right? Why would I want to do that?

Reply:
... but whatever.